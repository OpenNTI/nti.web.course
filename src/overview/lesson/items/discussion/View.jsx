import React from 'react';
import PropTypes from 'prop-types';
import {scoped} from '@nti/lib-locale';

import {Grid, List} from '../../Constants';
import Registry from '../Registry';

import ListCmp from './List';
import GridCmp from './Grid';

const DEFAULT_TEXT = {
	commentLabel: {
		topics: {
			one: '%(count)s Discussion',
			other: '%(count)s Discussions'
		},
		comments: {
			one: '%(count)s Comment',
			other: '%(count)s Comments'
		}
	}
};
const t = scoped('course.overview.lesson.items.discussion.View', DEFAULT_TEXT);


const HANDLES = [
	'application/vnd.nextthought.discussionref',
	'application/vnd.nextthought.discussion'
];

export default
@Registry.register(HANDLES)
class LessonOverviewDiscussion extends React.Component {
	static propTypes = {
		item: PropTypes.object,
		course: PropTypes.object,
		layout: PropTypes.oneOf([Grid, List])
	}

	state = {}


	componentDidMount () {
		this.setupFor(this.props);
	}


	componentDidUpdate (prevProps) {
		const {item:nextItem, course:nextCourse} = this.props;
		const {item:oldItem, course:oldCourse} = prevProps;

		if (nextItem !== oldItem || nextCourse !== oldCourse) {
			this.setupFor(this.props);
		}
	}


	componentWillUnmount () {
		this.unmounted = this;
		this.setState = () => {};
	}


	setupFor (props = this.props) {
		const {item, course} = this.props;

		this.setupTopic(item, course);
		this.setupIcon(item, course);
	}


	async setupTopic (item, course) {
		try {
			const topic = await item.resolveTarget(course);
			const isForum = topic.hasOwnProperty('TopicCount');

			this.setState({
				topic,
				title: topic.title,
				commentLabel: isForum ?
					t('commentLabel.topics', {count: topic.TopicCount || 0}) :
					t('commentLabel.comments', {count: topic.PostCount || 0})
			});
		} catch (e) {
			this.setState({
				disabled: true,
				title: item.title
			});
		}
	}


	async setupIcon (item, course) {
		if (!item.icon) { return; }

		try {
			const icon = await course.resolveContentURL(item.icon);

			this.setState({
				icon
			});
		} catch (e) {
			//No need to do anything here
		}
	}


	render () {
		const {layout, ...otherProps} = this.props;

		const Cmp = layout === List ? ListCmp : GridCmp;

		return (
			<Cmp layout={layout} {...otherProps} {...this.state} />
		);
	}
}
