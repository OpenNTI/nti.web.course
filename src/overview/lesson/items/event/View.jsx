import React from 'react';
import PropTypes from 'prop-types';
import {Models} from '@nti/lib-interfaces';

import PaddedContainer from '../../common/PaddedContainer';
import { List } from '../../Constants';
import Registry from '../Registry';

import BaseItem from './BaseItem';

export default
@Registry.register(Models.calendar.CalendarEventRef)
class LessonOverviewEvent extends React.Component {

	static propTypes = {
		course: PropTypes.object.isRequired,
		item: PropTypes.object.isRequired,
		layout: PropTypes.any
	}

	render () {
		const { layout, ...otherProps } = this.props;

		return (
			<PaddedContainer>
				<BaseItem {...otherProps} isMinimal={layout === List}/>
			</PaddedContainer>
		);
	}
}
