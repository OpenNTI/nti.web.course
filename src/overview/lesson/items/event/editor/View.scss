@import '~@nti/style-common';

.event-editor {
	background-color: white;

	.event-browser {
		.toolbar {
			height: 60px;
			position: relative;
			box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.15);

			.create-new {
				position: absolute;
				right: 10px;
				top: 10px;
				height: 40px;
				width: 40px;
				border-radius: 40px;
				color: white;
				background-color: $primary-blue;
				cursor: pointer;
				text-align: center;
				padding-top: 10px;
				box-shadow: 0 3px 11px 0 rgba(0, 0, 0, 0.3);
			}
		}

		.empty-state {
			.add-new {
				color: $primary-blue;
				cursor: pointer;
				margin-top: 10px;
			}
		}
	}
}
