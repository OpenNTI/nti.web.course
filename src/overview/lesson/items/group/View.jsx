import React from 'react';
import PropTypes from 'prop-types';

import {Grid, List} from '../../Constants';
import Registry from '../Registry';

import ListCmp from './List';
import GridCmp from './Grid';

@Registry.register('application/vnd.nextthought.nticourseoverviewgroup')
export default class LessonOverviewGroup extends React.Component {
	static propTypes = {
		layout: PropTypes.oneOf([Grid, List])
	}

	render () {
		const {layout, ...otherProps} = this.props;

		return layout === List ?
			(<ListCmp layout={layout} {...otherProps} />) :
			(<GridCmp layout={layout} {...otherProps} />);
	}
}
