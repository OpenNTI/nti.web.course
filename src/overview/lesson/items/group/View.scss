@import '~@nti/style-common';
@import '../../common/Variables.scss';

.nti-lesson-overview.grid-layout {
	.require-control-value {
		height: 28px;
		border-radius: 3px;
		padding: 0 6px;
		display: inline-block;

		span {
			height: 14px;
			line-height: 14px;
			font-size: 10px;
			color: $tertiary-grey;
			font-weight: bold;
		}

		i {
			color: $tertiary-grey;
		}
	}
}

.lesson-overview-group.grid-layout {
	.require-control-value {
		height: 28px;
		border-radius: 3px;
		padding: 0 6px;
		display: inline-block;

		span {
			height: 14px;
			line-height: 14px;
			font-size: 10px;
			color: $tertiary-grey;
			font-weight: bold;
		}

		i {
			color: $tertiary-grey;
		}
	}

	h2 {
		position: relative;
		font: normal 600 0.625rem/1.9 $body-font-family;
		text-transform: uppercase;
		color: white;
		display: inline-block;
		padding: 0 1rem 0 ($lesson-overview-desktop-padding + 1rem);

		&::before {
			content: '';
			display: block;
			position: absolute;
			top: 0;
			bottom: 0;
			left: 0;
			width: $lesson-overview-desktop-padding;
			background-color: rgba(255, 255, 255, 0.4);
		}

		@include responsive-range(tablet) {
			padding: 0 0.5rem 0 ($lesson-overview-tablet-padding + 0.5rem);

			&::before {
				width: $lesson-overview-tablet-padding;
			}
		}

		@include responsive-range(mobile) {
			padding: 0 0.5rem 0 ($lesson-overview-mobile-padding + 0.5rem);

			&::before {
				width: $lesson-overview-mobile-padding;
			}
		}
	}
}
