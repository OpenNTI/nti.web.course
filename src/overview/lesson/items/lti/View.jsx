import React from 'react';
import PropTypes from 'prop-types';

import { List, Grid } from '../../Constants';
import Registry from '../Registry';

import ListCmp from './List';
import GridCmp from './Grid';

export default
@Registry.register('application/vnd.nextthought.ltiexternaltoolasset')
class LTIExternalToolAsset extends React.Component {
	static propTypes = {
		layout: PropTypes.oneOf([Grid, List])
	}

	render () {
		const { layout, ...otherProps } = this.props;

		const Cmp = layout === List ? ListCmp : GridCmp;

		return (
			<Cmp layout={layout} {...otherProps} />
		);
	}
}
