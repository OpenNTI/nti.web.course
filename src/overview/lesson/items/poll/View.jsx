import React from 'react';

import Registry from '../Registry';

//TODO: figure out how to handle polls in the overview

export default class LessonOverviewPoll extends React.Component {
	render() {
		return null;
	}
}

Registry.register('application/vnd.nextthought.pollref')(LessonOverviewPoll);
