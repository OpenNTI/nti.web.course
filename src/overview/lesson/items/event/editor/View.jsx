import './View.scss';
import React from 'react';
import PropTypes from 'prop-types';
import { decorate } from '@nti/lib-commons';

import Overview from '../common/EventOverview';
import Browse from '../common/Browse';

import Store from './Store';

const BROWSE_LIST = 'browse-list';
const OVERVIEW = 'overview';

class EventEditor extends React.Component {
	static propTypes = {
		activePanel: PropTypes.string,
		course: PropTypes.object.isRequired,
		lessonOverview: PropTypes.object.isRequired,
		overviewGroup: PropTypes.object.isRequired,
		event: PropTypes.object,
		onCancel: PropTypes.func,
		onAddToLesson: PropTypes.func,
		onAddAsExternalLink: PropTypes.func,
		onDelete: PropTypes.func,
		saveDisabled: PropTypes.bool,
	};

	static deriveBindingFromProps({ course }) {
		return {
			course,
		};
	}

	state = {
		activePanel: BROWSE_LIST,
	};

	componentDidMount() {
		const { event, activePanel } = this.props;

		if (event) {
			this.setState({
				activePanel: OVERVIEW,
				event: event.CalendarEvent,
			});
		} else if (activePanel) {
			this.setState({ activePanel: activePanel });
		}
	}

	togglePanel = () => {
		if (this.state.activePanel === BROWSE_LIST) {
			this.setState({ activePanel: OVERVIEW });
		} else {
			this.setState({ activePanel: BROWSE_LIST });
		}
	};

	render() {
		const {
			course,
			onCancel,
			onAddToLesson,
			onDelete,
			lessonOverview,
			overviewGroup,
			saveDisabled,
			event: eventRef,
		} = this.props;
		const { event } = this.state;

		return (
			<div className="event-editor">
				{this.state.activePanel === BROWSE_LIST && (
					<Browse
						course={course}
						onSelect={selectedEvent => {
							this.setState({
								event: selectedEvent,
								activePanel: OVERVIEW,
							});
						}}
					/>
				)}
				{this.state.activePanel === OVERVIEW && (
					<Overview
						lessonOverview={lessonOverview}
						overviewGroup={overviewGroup}
						event={event}
						item={eventRef}
						course={course}
						onCancel={onCancel}
						onAddToLesson={onAddToLesson}
						onDelete={onDelete}
						saveDisabled={saveDisabled}
					/>
				)}
			</div>
		);
	}
}

export default decorate(EventEditor, [Store.connect(['loading', 'events'])]);
