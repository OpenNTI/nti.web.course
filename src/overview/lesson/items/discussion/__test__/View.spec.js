import React from 'react';
import renderer from 'react-test-renderer';
import { Summary } from '@nti/lib-interfaces';
import { TestUtils } from '@nti/web-client';

import View from '../View';
import { Grid } from '../../../Constants';

const { tearDownTestClient, setupTestClient } = TestUtils;

const getMockService = numberOfUsers => {
	return {
		async getObject() {
			return {
				title: 'Topic 1',
				TopicCount: 4,
				PostCount: 5,
			};
		},
	};
};

function itemBuilder() {
	return {
		_id: 'item1',
		_MimeType: 'application/vnd.nextthought.discussionref',
		_icon: 'http://www.test123.com/sample.png',
		_commentCount: null,
		_CompletionRequired: false,
		_title: 'Discussion item 1',
		_description: 'this is a test discussion overview item',
		_byline: 'Discussion Creator',
		_completable: false,

		commentCount: function (val) {
			this._commentCount = val;
			return this;
		},

		completable: function () {
			this._completable = true;
			return this;
		},

		required: function () {
			this._CompletionRequired = true;
			return this;
		},

		title: function (val) {
			this._title = val;
			return this;
		},

		description: function (val) {
			this._description = val;
			return this;
		},

		byline: function (val) {
			this._byline = val;
			return this;
		},

		build: function () {
			return {
				getID: () => this._id,
				icon: this._icon,
				MimeType: this._MimeType,
				[Summary]: {
					ItemCount: this._commentCount,
				},
				isCompletable: () => this._completable,
				CompletionRequired: this._CompletionRequired,
				title: this._title,
				description: this._description,
				byline: this._byline,
				resolveTarget: async () => {
					return {
						title: 'Topic 1',
						TopicCount: 4,
						PostCount: 5,
					};
				},
			};
		},
	};
}

/* eslint-env jest */
describe('Course overview discussion item test', () => {
	beforeEach(() => {
		jest.useFakeTimers();
		setupTestClient(getMockService());
	});

	afterEach(() => {
		tearDownTestClient();
	});

	test('Grid item with 5 comments, completable, required', async () => {
		const item = itemBuilder()
			.commentCount(5)
			.completable()
			.required()
			.build();

		const cmp = renderer.create(<View item={item} layout={Grid} />);

		const tree = cmp.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
