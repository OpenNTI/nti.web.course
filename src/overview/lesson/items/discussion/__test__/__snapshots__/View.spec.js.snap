// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course overview discussion item test Grid item with 5 comments, completable, required 1`] = `
<div
  className="nti-link-to-path lesson-overview-discussion-link"
  to="#"
>
  <div
    className="lesson-overview-discussion-grid-item"
  >
    <div
      className="icon default"
      style={null}
    />
    <div
      className="wrap"
    >
      <div
        className="title"
      >
        Discussion
      </div>
      <div
        className="comments"
      />
    </div>
  </div>
</div>
`;
